native/src/snappyclient/cpp/impl/ControlConnection.cpp

```diff
@@ -440,3 +449,35 @@ void ControlConnection::failoverExhausted(
   snappyEx.__set_serverInfo(failedServerString);
   throw snappyEx;
 }
+
+void ControlConnection::getConnectedHost(thrift::HostAddress& hostAddr,
+    thrift::HostAddress& connectedHost) {
+  boost::lock_guard<boost::mutex> controlConnLock(this->m_lock);
+
+  auto it = m_controlHostSet.find(hostAddr);
+  if (it != m_controlHostSet.end()){
+    connectedHost = it->second;
+    return;
+  }
+
+  for (auto iterator = m_controlHostSet.begin();
+      iterator != m_controlHostSet.end(); ++iterator) {
+    auto host = iterator->first;
+    if(host.__isset.isCurrent && host.isCurrent){
+      connectedHost = host;
+    }
+  }
+
+}
+
+void ControlConnection::close(bool clearGlobal) {
+  m_controlHost = thrift::HostAddress();
+  if (m_controlLocator != nullptr) {
+    m_controlLocator->getOutputProtocol()->getTransport()->close();
+    m_controlLocator.reset(nullptr);
+  }
+  if (clearGlobal) {
+    boost::lock_guard<boost::mutex> globalGuard(s_allConnsLock);
+    s_allConnections.erase(s_allConnections.begin());
```

Comment:
This is incorrect. What we need to remove is "this" from the list so need to find it first like:

```c++
  if (clearGlobal) {
    boost::lock_guard<boost::mutex> globalGuard(s_allConnsLock);
    auto thisPtr = std::unique_ptr<ControlConnection>(this);
    auto pos = std::find(s_allConnections.begin(), s_allConnections.end(), thisPtr);
    // don't delete "this" here
    thisPtr.release();
    if (pos != s_allConnections.end()) {
      s_allConnections.erase(pos);
    }
  }
```

Reply:
Modified as suggested